contrib/babelfishpg_tsql/src/properties.c

```diff
@@ -156,10 +164,102 @@ void* get_servername_helper()
     return info;
 }
 
+static void
+ProcessVersionNumber()
+{
+	int 		part = 0;
+	char 		*token;
+	const char 	*sql_server_version;
+	char		*copy_version_number = malloc(sizeof(sql_server_version));
+
+	sql_server_version = GetConfigOption("babelfishpg_tsql.version", true, false);
+	strcpy(copy_version_number,sql_server_version);
+	for (token = strtok(copy_version_number, "."); token; token = strtok(NULL, "."))
+	{ 
+		if(part == 0)
+			MajorVersion = atoi(token);
+		else if(part == 1)
+		{
+			MinorVersionDigits = strlen(token);
+			MinorVersion = atoi(token);
+		}
+		else
+			return;
+
+		part++; 
+	}
+}
+
+void* get_product_major_version_helper()
+{
+	StringInfoData	temp; 
+	void			*info;
+	const char 		*sql_server_version;
+	char	   		*major_version = (char *) palloc(3);	/* 2 digits, '\0' */
+
+	sql_server_version = GetConfigOption("babelfishpg_tsql.version", true, false);
+    initStringInfo(&temp);
+	if(strcasecmp(sql_server_version,"default") == 0)
+		appendStringInfoString(&temp, BABEL_COMPATIBILITY_MAJOR_VERSION);
+    else
+	{
+		pg_itoa(MajorVersion, major_version);
+		appendStringInfoString(&temp, major_version);
+	}
+		
+    info = tsql_varchar_input(temp.data, temp.len, -1);
+    pfree(temp.data);
+    return info;
+}
+
+void* get_product_minor_version_helper()
+{
+	StringInfoData	temp; 
+	void			*info;
+	char	   		*minor_version = (char *) palloc(MinorVersionDigits + 1);
+	const char 		*sql_server_version;
+	const char 		*ret = "0";
+
+	sql_server_version = GetConfigOption("babelfishpg_tsql.version", true, false);
+    initStringInfo(&temp);
+	if(strcasecmp(sql_server_version,"default") == 0)
+		appendStringInfoString(&temp, ret);
+    else
+	{
+		pg_itoa(MinorVersion, minor_version);
+		appendStringInfoString(&temp, minor_version);
+	}
+		
+    info = tsql_varchar_input(temp.data, temp.len, -1);
+    pfree(temp.data);
+    return info;
+}
+
+void* get_product_version_helper()
+{
+	StringInfoData	temp; 
+	void			*info;
+	const char 		*sql_server_version;
+	
+	sql_server_version = GetConfigOption("babelfishpg_tsql.version", true, false);
+
+    initStringInfo(&temp);
+	if(strcasecmp(sql_server_version,"default") == 0)
+		appendStringInfoString(&temp, BABEL_COMPATIBILITY_VERSION);
+    else
+		appendStringInfoString(&temp, sql_server_version);
+		
+    info = tsql_varchar_input(temp.data, temp.len, -1);
+    pfree(temp.data);
+    return info;
+}
+
 Datum serverproperty(PG_FUNCTION_ARGS) {
 	const char *property = text_to_cstring(PG_GETARG_TEXT_P(0));
 	VarChar *vch = NULL;
 	int64_t intVal = 0;
+
+	ProcessVersionNumber();
```

Comment:
This is only needed for 3 cases. Let's parse the version number only when it is needed.